### PR TITLE
Add support for Restart from Stage

### PR DESCRIPTION
**Pull request by chrishiner**
# PR Details

Allow Restart from Stage to work correctly

## Description

Per discussion in #197 
Updated TemplateFlowDefinition so that it calls the Actions on the Job, if any. 
Used this as an example:
[https://github.com/jenkinsci/workflow-cps-plugin/blob/master/src/main/java/org/jenkinsci/plugins/workflow/cps/CpsScmFlowDefinition.java#L102
](url)

## How Has This Been Tested

Run build without restart from stage with stash in early stage.
Run build with restart from stage at a later stage, verified that stash was found.
Restart Jenkins.
Run build with restart from stage at a later stage, verified that stash was found.

Jenkins 2.277.2 running in Kubernetes.

## Types of Changes

- [x] Added Unit Testing
- [ ] Docs change / refactoring / dependency upgrade
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [ ] I have added the appropriate label for this PR
- [ ] If necessary, I have updated the documentation accordingly.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by cokieffebah** [comment]
Thanks. will review in next few days

File: src/main/groovy/org/boozallen/plugins/jte/job/TemplateFlowDefinition.groovy [hunk 1]
```diff
@@ -52,6 +53,14 @@ abstract class TemplateFlowDefinition extends FlowDefinition {
     FlowExecution create(FlowExecutionOwner owner, TaskListener listener, List<? extends Action> actions) throws Exception {
         FlowDurabilityHint hint = determineFlowDurabilityHint(owner)
         String template = initializeJTE(owner)
+
```
Comment (cokieffebah) [inline src/main/groovy/org/boozallen/plugins/jte/job/TemplateFlowDefinition.groovy:56]: The action usage should happen before the initializeJTE call. basically put the action loop at the top

**Review by cokieffebah — CHANGES_REQUESTED** [review]

**Comment by cokieffebah** [comment]
You should look at the Pipeline Definition code and add a/some unit tests that try to replicate their unit test(s) of stage restarts.

File: src/main/groovy/org/boozallen/plugins/jte/job/TemplateFlowDefinition.groovy (hunk 1 above)
Comment (chrishiner) [reply src/main/groovy/org/boozallen/plugins/jte/job/TemplateFlowDefinition.groovy:56]: I tried that, but then it failed to find any of the JTE variables when it ran. The hint part can probably move after the actions, but the initializeJTE part needs to happen first.

Comment (cokieffebah) [reply src/main/groovy/org/boozallen/plugins/jte/job/TemplateFlowDefinition.groovy:56]: ok. let me get some feedback from Steven

**Comment by chrishiner** [comment]
I almost have a working unit test, but I can't seem to get it to create the workflow job in a way that works both for JTE and for a pipeline. I'm sure I'm missing something simple.
If I do:
```
        def pipelinetemplate="pipeline......"
        WorkflowJob p = TestUtil.createAdHoc(
            config: '',
            template: pipelinetemplate.
        jenkins, 'p')
```
Then it fails because it's not being processed as a pipeline.
If I add:
        p.setDefinition(new CpsFlowDefinition(pipelinetemplate, true))
then it does the pipeline parts, but JTE doesn't get enabled.

**Comment by cokieffebah** [comment]
@chrishiner trying to see if I can get your test working without the resume having to be in the library.  not successful yet.  will take another look tomorrow US Eastern time

**Comment by cokieffebah** [comment]
Note to self: Look into https://github.com/jenkinsci/workflow-api-plugin/blob/master/src/main/java/org/jenkinsci/plugins/workflow/flow/FlowCopier.java

**Comment by cokieffebah** [comment]
reviewing now. @chrishiner

File: src/test/groovy/org/boozallen/plugins/jte/init/RestartFromStageSpec.groovy [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+*/
+package org.boozallen.plugins.jte.init
+
+import com.gargoylesoftware.htmlunit.html.HtmlForm
+import com.gargoylesoftware.htmlunit.html.HtmlPage
+import com.gargoylesoftware.htmlunit.html.HtmlSelect
```
Comment (cokieffebah) [inline src/test/groovy/org/boozallen/plugins/jte/init/RestartFromStageSpec.groovy:20]:
```suggestion
```
I realize we put these in, but could you remove these imports
```
import com.gargoylesoftware.htmlunit.html.HtmlForm
import com.gargoylesoftware.htmlunit.html.HtmlPage
import com.gargoylesoftware.htmlunit.html.HtmlSelect
```

File: src/test/groovy/org/boozallen/plugins/jte/init/RestartFromStageSpec.groovy [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+import hudson.model.Action
+import hudson.model.Cause
+import hudson.model.CauseAction
+import org.boozallen.plugins.jte.init.governance.libs.TestLibraryProvider
+import org.boozallen.plugins.jte.util.TestUtil
+import org.jenkinsci.plugins.pipeline.modeldefinition.actions.RestartDeclarativePipelineAction
```
Comment (cokieffebah) [inline src/test/groovy/org/boozallen/plugins/jte/init/RestartFromStageSpec.groovy:26]:
```suggestion
```
and this one.
```
import org.jenkinsci.plugins.pipeline.modeldefinition.actions.RestartDeclarativePipelineAction
```

**Review by cokieffebah — CHANGES_REQUESTED** [review]

File: src/test/groovy/org/boozallen/plugins/jte/init/RestartFromStageSpec.groovy (hunk 2 above)
Comment (steven-terrana) [reply src/test/groovy/org/boozallen/plugins/jte/init/RestartFromStageSpec.groovy:20]: just opened https://github.com/chrishiner/templating-engine-plugin/pull/3 to resolve

File: src/test/groovy/org/boozallen/plugins/jte/init/RestartFromStageSpec.groovy (hunk 3 above)
Comment (steven-terrana) [reply src/test/groovy/org/boozallen/plugins/jte/init/RestartFromStageSpec.groovy:26]: just opened https://github.com/chrishiner/templating-engine-plugin/pull/3 to resolve

**Review by cokieffebah — APPROVED** [review]

**Pull request merged**